Std/Tactic/Simps/Basic.lean

```diff
@@ -0,0 +1,1157 @@
+    if cfg.fullyApplied then
+      addProjection stxProj univs nm tgt lhsAp rhsAp newArgs cfg
+    else
+      addProjection stxProj univs nm type lhs rhs args cfg
+    return #[nm]
+  -- if the type is a structure
+  let some (.inductInfo { isRec := false, ctors := [ctor], .. }) := env.find? str | unreachable!
+  trace[simps.debug] "{str} is a structure with constructor {ctor}."
+  let rhsEta ← headStructureEtaReduce rhsAp
+  -- did the user ask to add this projection?
+  let addThisProjection := stx?.isSome && toApply.isEmpty
+  if addThisProjection then
+    -- we pass the precise argument of simps as syntax argument to `addProjection`
+    if cfg.fullyApplied then
+      addProjection stxProj univs nm tgt lhsAp rhsEta newArgs cfg
+    else
+      addProjection stxProj univs nm type lhs rhs args cfg
+  let rhsWhnf ← withTransparency cfg.rhsMd <| whnf rhsEta
+  trace[simps.debug] "The right-hand-side {indentExpr rhsAp}\n reduces to {indentExpr rhsWhnf}"
+  if !rhsWhnf.getAppFn.isConstOf ctor then
+    -- if I'm about to run into an error, try to set the transparency for `rhsMd` higher.
+    if cfg.rhsMd == .reducible && (mustBeStr || !todoNext.isEmpty || !toApply.isEmpty) then
+      trace[simps.debug] "Using relaxed reducibility."
+      Linter.logLintIf linter.simpsNoConstructor ref
+        m!"The definition {nm} is not a constructor application. Please use `@[simps!]` instead.{
+        ""}\n\nExplanation: `@[simps]` uses the definition to find what the simp lemmas should {
+        ""}be. If the definition is a constructor, then this is easy, since the values of the {
+        ""}projections are just the arguments to the constructor. If the definition is not a {
+        ""}constructor, then `@[simps]` will unfold the right-hand side until it has found a {
+        ""}constructor application, and uses those values.\n\n{
+        ""}This might not always result in the simp-lemmas you want, so you are advised to use {
+        ""}`@[simps?]` to double-check whether `@[simps]` generated satisfactory lemmas.\n{
+        ""}Note 1: `@[simps!]` also calls the `simp` tactic, and this can be expensive in certain {
+        ""}cases.\n{
+        ""}Note 2: `@[simps!]` is equivalent to `@[simps (config := \{rhsMd := .default, {
+        ""}simpRhs := true})]`. You can also try `@[simps (config := \{rhsMd := .default})]` {
+        ""}to still unfold the definitions, but avoid calling `simp` on the resulting statement.\n{
+        ""}Note 3: You need `simps!` if not all fields are given explicitly in this definition, {
+        ""}even if the definition is a constructor application. For example, if you give a {
+        ""}`MulEquiv` by giving the corresponding `Equiv` and the proof that it respects {
+        ""}multiplication, then you need to mark it as `@[simps!]`, since the attribute needs to {
+        ""}unfold the corresponding `Equiv` to get to the `toFun` field."
```

Comment:
I'd love to see this function and the one above broken up into smaller entities with comments about what they do.

Reply:
Yeah, this code is not pretty, since I wrote it in Lean 3, and then ported it to Lean 4 keeping the Lean 3 structure. I think the proper way to make this code better is to make use of monad extensions with a state and context, instead of these functions that take ~10 arguments. That makes it also much easier to split off small sub-functions.
I don't have time for that refactor at the moment.

It is a bit tricky to break up this function in the current state, since there are recursive calls in multiple places.

Reply:
Thanks.

While it would be really nice to get `@[simps]` into Std, it is not essential: users can always write the boilerplate by hand!

I propose that we shop this project around a bit, and see if anyone on zulip might be interested in taking on a refactor?